### PR TITLE
Added `databricks_service_principal` and `databricks_service_principals` data resources

### LINKED ISSUE #1208: [FEATURE] Add data source for SPNs and identify if current user is a SPN
Summary says it all...

I have a couple of uses cases where it's helpful to identify if the script is being run under the scope of a service principal (easily facilitated by checking if the current user has an application id), as well as being able to query for an existing service principal.

I've submitted a PR to this effect #1207 . It's my first time writing go, so please go easy on me :)

### PR DESCRIPTION
**Pull request by nkvuong**
This continues and closes #1207 

- return `application_id` from `databricks_current_user` if that user is a Service Principal
- add `databricks_service_principal` data source based on `application_id`
- add `databricks_service_principals` data source based on `display_name`
- add tests for data sources
- update documentation
Closes #1208

### CODE REVIEW

File: docs/data-sources/service_principal.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+- `application_id` - (Required) ID of the service principal. The service principal must exist before this resource can be retrieved.
+
+## Attribute Reference
+
+Data source exposes the following attributes:
+
+- `sp_id` - The id of the service principal.
+- `application_id` - This is the application id of the given service principal
+- `external_id` - ID of the service principal in an external identity provider.
```
Comment (nfx) [inline docs/data-sources/service_principal.md:42]: Isn't it the same as application id?

File: scim/data_service_principal.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"context"
+	"fmt"
+
+	"github.com/databrickslabs/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// DataSourceServicePrincipal returns information about the spn specified by the application_id
+func DataSourceServicePrincipal() *schema.Resource {
+	type spnData struct {
```
Comment (nfx) [inline scim/data_service_principal.go:13]: Can we reuse the entity from SPN resource, by the way?

File: scim/data_service_principal.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		response.Repos = fmt.Sprintf("/Repos/%s", sp.ApplicationID)
+		response.ExternalID = sp.ExternalID
+		response.Active = sp.Active
+		response.SpID = sp.ID
+		return nil
+	})
+}
+
+// DataSourceServicePrincipal returns information about the spn specified by the application_id
+func DataSourceServicePrincipals() *schema.Resource {
```
Comment (nfx) [inline scim/data_service_principal.go:44]: Can you put them into separate files?

File: scim/data_service_principal.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		spnAPI := NewServicePrincipalsAPI(ctx, c)
+		spList, err := spnAPI.filter(fmt.Sprintf("displayName eq '%s'", response.DisplayName))
+		if err != nil {
+			return err
+		}
+		if len(spList) == 0 {
+			return fmt.Errorf("cannot find SPs with display name %s", response.DisplayName)
+		}
+		for _, sp := range spList {
+			response.ApplicationIDs = append(response.ApplicationIDs, sp.ApplicationID)
```
Comment (nfx) [inline scim/data_service_principal.go:60]: Don't forget to sort everything that is a set

File: scim/data_user.go [hunk 5]
```diff
@@ -71,6 +75,7 @@ func DataSourceUser() *schema.Resource {
 			d.Set("home", fmt.Sprintf("/Users/%s", user.UserName))
 			d.Set("repos", fmt.Sprintf("/Repos/%s", user.UserName))
 			d.Set("external_id", user.ExternalID)
+			d.Set("application_id", user.ApplicationID)
```
Comment (nfx) [inline scim/data_user.go:78]: Why?

File: scim/data_user.go [hunk 6]
```diff
@@ -59,6 +59,10 @@ func DataSourceUser() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+			"application_id": {
```
Comment (nfx) [inline scim/data_user.go:62]: Why?

File: docs/data-sources/service_principal.md (hunk 1 above)
Comment (nkvuong) [reply docs/data-sources/service_principal.md:42]: they are different - application_id is a uuid but the id is a long

File: scim/data_service_principal.go (hunk 2 above)
Comment (nkvuong) [reply scim/data_service_principal.go:13]: this returns a few more computed fields

File: docs/data-sources/service_principals.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  group_id  = data.databricks_group.admins.id
+  member_id = data.databricks_service_principal.spn[each.value].sp_id
+}
+```
+
+## Argument Reference
+
+Data source allows you to pick service principals by the following attributes
+
+- `display_name` - (Required) Display name of the service principals. The service principals must exist before this resource can be retrieved.
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply docs/data-sources/service_principals.md:40]: rename it to contains

File: scim/data_service_principal.go (hunk 3 above)
Comment (nkvuong) [reply scim/data_service_principal.go:44]: yes

File: scim/data_user.go (hunk 5 above)
Comment (nkvuong) [reply scim/data_user.go:78]: if databricks_current_user  is a Service Principal, return the application_id as well

File: scim/data_service_principal.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+// DataSourceServicePrincipal returns information about the spn specified by the application_id
+func DataSourceServicePrincipals() *schema.Resource {
+	type spnsData struct {
+		DisplayName    string   `json:"display_name,omitempty" tf:"computed"`
+		ApplicationIDs []string `json:"application_ids,omitempty" tf:"computed,slice_set"`
+	}
+	return common.DataResource(spnsData{}, func(ctx context.Context, e interface{}, c *common.DatabricksClient) error {
+		response := e.(*spnsData)
+		spnAPI := NewServicePrincipalsAPI(ctx, c)
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply scim/data_service_principal.go:51]: this is useful for looking up spns either created outside of Terraform, or in another Terraform config - application_id or id are less friendly

File: docs/data-sources/service_principal.md (hunk 1 above)
Comment (nfx) [reply docs/data-sources/service_principal.md:41]: I actually meant the external id, hard to review from phone

**Review by nfx — APPROVED** [review]
🎊

**Pull request merged**